### PR TITLE
Implement display name resolving

### PR DESCRIPTION
**Pull request by vbrandl**
As discussed with @poljar in https://github.com/poljar/weechat-matrix-rs/pull/11, this tries to resolve the username on each membership change and exposes a function to get the resolved username for a room member.

If there are two members with the same display name, the name is resolved to `display_name (user_id)`.

This is a draft because I'm not quite happy about the amount of cloning that is performed but it's the best I can come up with for now. But maybe anyone has an idea how to get rid of some of the clones.

### CODE REVIEW

File: matrix_sdk_base/src/models/room.rs [hunk 1]
```diff
@@ -257,6 +265,63 @@ impl Room {
...
+                    .as_ref()
+                    .map(|n| &n == v)
+                    .unwrap_or(false)
+            })
+            .map(|(k, _)| k)
+            .cloned()
+            .collect();
+
+        // if there is no other user with the same display name -> just use the display name
+        if users_with_same_name.is_empty() {
```
Comment (poljar) [inline matrix_sdk_base/src/models/room.rs:284]: Can't we store just the users that have clashes? That way you wouldn't need to duplicate the display names of every user in the room and will get rid of the majority of clones that are happening.

File: matrix_sdk_base/src/models/room.rs [hunk 2]
```diff
@@ -247,6 +250,11 @@ impl Room {
         self.encrypted
     }
 
+    /// Get the resolved display name for a member of this room.
+    pub fn resolved_display_name(&self, id: &UserId) -> Option<&str> {
```
Comment (poljar) [inline matrix_sdk_base/src/models/room.rs:254]: Can we name this to be something like `user_display_name()` or `member_display_name()`?

Comment (vbrandl) [reply matrix_sdk_base/src/models/room.rs:254]: And drop the `resolve` part?

File: matrix_sdk_base/src/models/room.rs (hunk 1 above)
Comment (vbrandl) [reply matrix_sdk_base/src/models/room.rs:284]: That's a good idea. Also I just realised, I forgot to cleanup the names each time a user leaves the room.

File: matrix_sdk_base/src/models/room.rs (hunk 2 above)
Comment (poljar) [reply matrix_sdk_base/src/models/room.rs:254]: Yeah.

File: matrix_sdk_base/src/models/room.rs (hunk 1 above)
Comment (poljar) [reply matrix_sdk_base/src/models/room.rs:284]: Also if a user changes their display name.

**Comment by DevinR528** [comment]
If you wanted to use this instead then it wouldn't be quite as touchy (formating and spacing matters and is a pain)

<details>
<summary>using serde_json::json! instead of just String</summary>

```rust
#[cfg(not(feature = "messages"))]
assert_eq!(
    serde_json::json!({
        "!roomid:example.com": {
            "room_id": "!roomid:example.com",
            "room_name": {
                "name": null,
                "canonical_alias": null,
                "aliases": [],
                "heroes": [],
                "joined_member_count": null,
                "invited_member_count": null
            },
            "own_user_id": "@example:example.com",
            "creator": null,
            "members": {},
            "typing_users": [],
            "power_levels": null,
            "encrypted": false,
            "unread_highlight": null,
            "unread_notifications": null,
            "tombstone": null
        }
    }),
    serde_json::to_value(&joined_rooms).unwrap()
);

#[cfg(feature = "messages")]
assert_eq!(
    serde_json::json!({
        "!roomid:example.com": {
            "room_id": "!roomid:example.com",
            "room_name": {
                "name": null,
                "canonical_alias": null,
                "aliases": [],
                "heroes": [],
                "joined_member_count": null,
                "invited_member_count": null
            },
            "own_user_id": "@example:example.com",
            "creator": null,
            "members": {},
            "messages": [],
            "typing_users": [],
            "power_levels": null,
            "encrypted": false,
            "unread_highlight": null,
            "unread_notifications": null,
            "tombstone": null
        }
    }),
    serde_json::to_value(&joined_rooms).unwrap()
);
```
</details>

File: matrix_sdk_base/src/models/room.rs (hunk 2 above)
Comment (vbrandl) [reply matrix_sdk_base/src/models/room.rs]: Done in b98d2388db0129d80f3f238c0e8632be93bd6929

**Comment by vbrandl** [comment]
Failing test was fixed in 25b941308e799d007a13890da9051d80007d5013

**Comment by vbrandl** [comment]
@poljar: I added the following test: https://github.com/matrix-org/matrix-rust-sdk/blob/2214b50c20403670052b0a9c07825f6864e7122f/matrix_sdk_base/src/models/room.rs#L680

But in the end there is always only one user in the room. How can I simulate server state in the unit tests? I need a test server that holds the state (e.g. members per room).

Or is it enough to create a `Room` object and call `add_member` and `update_member` a bunch of times?

**Comment by poljar** [comment]
> @poljar: I added the following test:
> 
> https://github.com/matrix-org/matrix-rust-sdk/blob/2214b50c20403670052b0a9c07825f6864e7122f/matrix_sdk_base/src/models/room.rs#L680
> 
> But in the end there is always only one user in the room. How can I simulate server state in the unit tests? I need a test server that holds the state (e.g. members per room).
> 

You are creating two clients, both of them receive the same sync response. You'll likely want to have one client receiving two different sync responses. You can add custom events to a joined room with [this method](https://github.com/matrix-org/matrix-rust-sdk/blob/master/matrix_sdk_test/src/lib.rs#L128) of the EventBuilder.

> Or is it enough to create a `Room` object and call `add_member` and `update_member` a bunch of times?

It is enough, as long as you are hitting the new code-paths that you added. The `receive_sync_method()` does exactly that after it finds an event that changes the membership of a  room. In both cases you'll likely create some custom `m.room.member` events that change the room state.

**Comment by poljar** [comment]
This has been superseded by https://github.com/matrix-org/matrix-rust-sdk/pull/55 which has been merged.

**Pull request closed without merging**